### PR TITLE
Adopt ansible-compat runtime

### PR DESCRIPTION
**Pull request by ssbarnea**
As part of decoupling molecule from ansible-lint, the functionality from ansiblelint.prerun was moved in a standalone generic library named ansible-compat.

This change is marked as major because it will break compatibility for anyone that relied on use of prerun module from the linter. The effective removal of prerun submodule will be done
in a follow-up, just to allow for an easier review.

Fixes: #1712
Fixes: #1784

### CODE REVIEW

File: setup.cfg [hunk 1]
```diff
@@ -71,6 +71,7 @@ install_requires =
   # ansible-lint does not list ansible as direct dependency in order to
   # allow user to choose his target version. Be sure that you mention one
   # of the extras or install a specific version of ansible yourself.
+  ansible-compat>=0.3.0
```
Comment (felixfontein) [inline setup.cfg:74]: I really think we should stick to a stable version, and not an initial development version which can easily have breaking changes.
```suggestion
  ansible-compat>=1.0.0,<2.0.0
```

**Review by felixfontein — CHANGES_REQUESTED** [review]

File: setup.cfg (hunk 1 above)
Comment (webknjaz) [reply setup.cfg:74]: For 0.x it's usually capped with the next minor:
```suggestion
  ansible-compat >=0.3.0, <0.4.0
```
No need to force a major release just yet — we have a perfect opportunity to iterate on the API and locking it up at this point is a bad idea.
Having the same people in charge of the both projects makes it rather safe. OTOH it may be reasonable to have some downstream tests integrated into the compat's CI to be on the safe side.
Also, zerover does not mean development. Dev versions have suffices like `.dev0`. 0ver just means that the project is not yet fully settled on the API which is not necessarily a bad thing.

**Comment by webknjaz** [comment]
Do you want to add `Closes #1712`?

File: setup.cfg [hunk 2]
```diff
@@ -71,6 +71,7 @@ install_requires =
   # ansible-lint does not list ansible as direct dependency in order to
   # allow user to choose his target version. Be sure that you mention one
   # of the extras or install a specific version of ansible yourself.
+  ansible-compat >=0.3.0, <0.4.0
```
Comment (webknjaz) [inline setup.cfg:74]: Let's try this, it's equivalent to the dep spec above but that is semantically closer to SemVer expectations:
```suggestion
  ansible-compat ~= 0.3.0
```

Comment (ssbarnea) [reply setup.cfg:74]: In fact I will completely remove the capping and we will add it only when we will break the API with compat. We know very well who are to be the consumers. That is because capping it too soon it would require extra maintenance for us.

File: src/ansiblelint/testing/__init__.py [hunk 3]
```diff
@@ -18,7 +18,8 @@
 
 # Emulate command line execution initialization as without it Ansible module
 # would be loaded with incomplete module/role/collection list.
-prepare_environment()
+runtime = Runtime(isolated=True)
+runtime.prepare_environment()
```
Comment (webknjaz) [inline src/ansiblelint/testing/__init__.py:22]: I don't really like this pattern of initializing a single behaviour with two separate things/invocations which must always be one after another with no code causing side-effects in between.

Comment (ssbarnea) [reply src/ansiblelint/testing/__init__.py:22]: You will have to get used to it because I remember @tadeboro suggesting months ago that doing extra optional stuff inside constructor was a bad idea. The two step process is there by design and not accident, so we have a simple constructor.

File: src/ansiblelint/rules/__init__.py [hunk 4]
```diff
@@ -39,6 +40,10 @@
 
 
 class AnsibleLintRule(BaseRule):
+    """Based class used by most rules."""
```
Comment (cognifloyd) [inline src/ansiblelint/rules/__init__.py:43]:
```suggestion
    """Base class used by most rules."""
```

File: src/ansiblelint/config.py [hunk 5]
```diff
@@ -60,6 +60,7 @@
     cwd=".",
     display_relative_path=True,
     exclude_paths=[],
+    format="rich",
```
Comment (cognifloyd) [inline src/ansiblelint/config.py:63]: What does this have to do with removing prerun?

File: src/ansiblelint/testing/__init__.py [hunk 6]
```diff
@@ -7,7 +7,10 @@
 import tempfile
 from typing import TYPE_CHECKING, Any, Dict, List, Optional
 
-from ansiblelint.prerun import prepare_environment
+from ansible_compat.runtime import Runtime
```
Comment (cognifloyd) [inline src/ansiblelint/testing/__init__.py:10]: Unused import?

File: src/ansiblelint/testing/__init__.py [hunk 7]
```diff
@@ -7,7 +7,10 @@
...
+
+from ansiblelint.app import get_app
+from ansiblelint.rules import RulesCollection  # noqa: E402
```
Comment (cognifloyd) [inline src/ansiblelint/testing/__init__.py:13]: I believe you can remove the noqa comment since this moved up, right?
```suggestion
from ansiblelint.rules import RulesCollection
```

**Review by cognifloyd — COMMENTED** [review]
Looking awesome! Just a few nitpicky comments/questions.

File: src/ansiblelint/config.py (hunk 5 above)
Comment (ssbarnea) [reply src/ansiblelint/config.py:63]: It is needed, otherwise I will get a runtime errors when I try to access .format in main. In fact I am not sure how that it worked before this but I suspect it has something to do on config initialization order.

**Review by cognifloyd — APPROVED** [review]
